framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,656 @@
+    }
+
+    /**
+     * Returns the part of a binary name that specifies the basename of the class.
+     *
+     * @param className a binary name
+     * @return the part of the name representing the class's name without its package
+     */
+    // A binary name without its package is still a binary name
+    @SuppressWarnings("signature:return.type.incompatible")
+    private static @BinaryName String basenamePart(@BinaryName String className) {
+        int lastdot = className.lastIndexOf('.');
+        return (lastdot == -1) ? className : className.substring(lastdot + 1);
+    }
+
+    /**
+     * Returns the String representation of an annotation in Java source format.
+     *
+     * @param a the annotation to print
+     * @return the formatted annotation
+     */
+    public static String formatAnnotation(Annotation a) {
+        String annoName = a.def().name.substring(a.def().name.lastIndexOf('.') + 1);
+        if (a.fieldValues.isEmpty()) {
+            return "@" + annoName;
+        }
+        StringJoiner sj = new StringJoiner(",", "@" + annoName + "(", ")");
+        for (Map.Entry<String, Object> f : a.fieldValues.entrySet()) {
+            AnnotationFieldType aft = a.def().fieldTypes.get(f.getKey());
+            sj.add(f.getKey() + "=" + IndexFileWriter.formatAnnotationValue(aft, f.getValue()));
+        }
+        return sj.toString();
+    }
+
+    /**
+     * Returns all annotations in {@code annos} in a form suitable to be printed as Java source
+     * code.
+     *
+     * <p>Each annotation is followed by a space, to separate it from following Java code.
+     *
+     * @param annos the annotations to format
+     * @return all annotations in {@code annos}, each followed by a space, in a form suitable to be
+     *     printed as Java source code
+     */
+    private static String formatAnnotations(Collection<? extends Annotation> annos) {
+        StringBuilder sb = new StringBuilder();
+        for (Annotation tla : annos) {
+            if (!isInternalJDKAnnotation(tla.def.name)) {
+                sb.append(formatAnnotation(tla));
+                sb.append(" ");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Formats the component types of an array via recursive descent through the array's scene-lib
+     * structure.
+     *
+     * @param e the array's scenelib type element
+     * @param arrayType the string representation of the array's type
+     * @return the type formatted to be written to Java source code, followed by a space character
+     */
+    private static String formatArrayType(ATypeElement e, String arrayType) {
+        StringBuilder result = new StringBuilder();
+        // If there are not annotations on the component, this will be true - unless there
+        // are also no annotations on the first array level. See comment below for how that
+        // case is handled.
+        int componentEndPos;
+        {
```

Comment:
Consider abstracting out this code to make `formatArrayType` shorter and simpler.
(The curly braces are intentional in case you don't do that, to clarify that the purpose of the contained code is just to set `componentEndPos`.  But I think abstracting it out would be better.)